consensus/scheme/rolldpos/rolldposctx.go

```diff
@@ -454,9 +487,16 @@ func (ctx *rollDPoSCtx) NewProposalEndorsement(msg interface{}) (interface{}, er
 		if err := ctx.round.AddBlock(proposal.block); err != nil {
 			return nil, err
 		}
+		if err := ctx.prepareNextProposal(proposal.block.Height(), blkHash); err != nil {
+			ctx.loggerWithStats().Warn("failed to prepare next proposal", zap.Error(err), zap.Uint64("prevHeight", proposal.block.Height()))
+		}
 		ctx.loggerWithStats().Debug("accept block proposal", log.Hex("block", blockHash))
 	} else if ctx.round.IsLocked() {
 		blockHash = ctx.round.HashOfBlockInLock()
+	} else {
+		if err := ctx.prepareNextProposal(ctx.round.Height()-1, ctx.round.PrevHash()); err != nil {
```

Comment:
why now need to `prepareNextProposal()` in this case?

Reply:
msg == nil means that the proposal for this round was not received in time; in this case, it's possible to mint the proposal for round +1 in advance.